onnxruntime/core/session/inference_session.cc

```diff
@@ -1864,6 +1873,75 @@ AllocatorPtr InferenceSession::GetAllocator(const OrtMemoryInfo& mem_info) const
   return session_state_->GetAllocator(mem_info);
 }
 
+void InferenceSession::ShrinkDefaultMemoryArenas(const std::string& ort_device_list) {
```

Comment:
This function is doing 2 things - parsing the string and invoking Shrink. If the string is invalid, we don't have to wait for Run() to finish before ignoring/warning the user. Essentially, the string parsing and validation can be done upfront and if it's invalid an error can be returned right away.

Reply:
done